[DRAFT] Mip-Splatting implementation

niujinshuchong opened the pull request:
Implementation of [Mip-Splatting](https://github.com/autonomousvision/mip-splatting). Tested on bicycle scenes and works OK.

Here is a comparison when trained with factor 4 and render with factor 1.


Needs more testing and clean up later.

jb-ye commented on examples/simple_trainer_mip_splatting.py:
```diff
@@ -0,0 +1,1069 @@
+        
+        scales_after_square = scales_square + torch.square(filters) [:, None] # [N, 1]
+        det2 = scales_after_square.prod(dim=1) # [N,]
+        coef = torch.sqrt(det1 / det2) # [N,]
```
Could the scale be very small, sqrt is not differentiable near zero. Maybe consider add a clamp to avoid zero differentiation?

liruilong940607 commented:
Awesome @niujinshuchong ! 

I'm thinking if there is a way to elegantly message some of implementation into the gsplat library, e.g., into `rasterization()` function so that the downstream users (e.g., nerfstudio) can just import and use.

It doesn't have to be living in the `rasterization()` function if it does not fit there. We could also put some implementation in to like `gsplat/mip.py`, so that ppl can use it with:
```
from gsplat.mip import ...
# and do their thing
```

liruilong940607 commented:
I'll come back to it later with a closer look. Great job!

niujinshuchong commented:
@liruilong940607 Thanks for the suggestions. I added a cuda implementation of compute_3D_smoothing_filter and it can be imported from gsplat. The benchmark results (see top) are reasonable. For now, I create a new script `simple_trainer_mip_splatting.py` and it almost the same as `simple_trainer.py`, do you think we should merge them or make it separate? I also added a training script to use multiple gpus.

liruilong940607 commented on examples/simple_trainer_mip_splatting.py:
```diff
@@ -0,0 +1,1094 @@
+        self.splats["filters"] = torch.nn.Parameter(filter_3D)
+
+    @torch.no_grad()
+    def compute_3D_smoothing_filter_torch(self):
```
Would you mind move this into `gsplat/cuda/_torch_impl.py` and put up a test case under `test/test_basic.py` to protect the CUDA implementation `compute_3D_smoothing_filter()`?

liruilong940607 commented:
Thanks for the benchmarking! Really nice! Yeah from a quick going through of the training script, it seems to be very compatible with the original one so I would suggest we merge the two! Just need a flag in the config, e.g.(--mip_splatting?) and a couple if else in the code.

liruilong940607 commented on gsplat/cuda/_wrapper.py:
```diff
@@ -572,6 +572,34 @@ def rasterize_to_indices_in_range(
     return out_gauss_ids, out_pixel_ids, out_camera_ids
 
 
+def compute_3D_smoothing_filter(
```
nit: maybe use `3d` instead of `3D` in the naming of these functions to align with the naming convention in gsplat?

liruilong940607 commented on examples/benchmark_mipnerf360.py:
Nice scripts! 
- Does `GPUtil` need to be installed via pip? If so we should put it into `examples/requirements.txt`.
- If you are going to merge the `simple_trainer_mip_splatting.py` with `simple_trainer.py` then we can just replace the original benchmark script with this one, and enable an input flag to enable mip splatting. But merging the benchmark script means that we probably want the default benchmarking set to be the original 3DGS.

liruilong940607 commented on examples/benchmark_mipnerf360_stmt.py:
```diff
@@ -0,0 +1,115 @@
+excluded_gpus = set([])
+
+# classic
+result_dir = "results/benchmark_stmt"
```
what does `stmt` stands for? lol

maturk replied:
I think` Single-scale-Training-and-Multi-scale-Testing`, the idea being you train with a fixed resolution (here a factor of 1/8 of the original image res) and test on various other resolutions (1x full res,  1/2, 1/4, 1/8).

liruilong940607 commented on examples/benchmark_mipnerf360_stmt.py:
how hard would it be to merge this one with the other script? Ideally an list argument could be passed in to specify the rendering factors that the evaluation happens on

maturk replied:
I think we could adapt the current `benchmark.sh` script into the `.py` style similar to the one proposed here, allowing for easier batch jobs on e.g. clusters and also streamlining the evaluation of the various new features of gsplat... I can look into it...

liruilong940607 commented on examples/show_mipnerf360.py:
Nice script! Would love to have it! Maybe the name of this script could be something like "compare_results_mipnerf360" to be more readable? Actually maybe we could also get rid of the `_mipnerf360` suffix until when another dataset is supported.

liruilong940607 commented on examples/show_mipnerf360_allscales.py:
```diff
@@ -0,0 +1,45 @@
+import json
```
actually, another idea of managing these scripts is to have a `mip_splatting` folder under `examples`, in which you could have multiple scripts in there, without worrying about merging or conflicting with other scripts. I think this is a scalable way to manage the scripts when more things are supported in the future.

And you could also have a customized `simple_trainer.py` living in it, basically whatever you need to be customized. But this is less preferred because a customized `simple_trainer.py` means the mip-splatting feature would be disjoint with other future features for user to play with. So ideally we merge the trainer but could put the scripts into an isolated folder.

jb-ye commented:
What's the plan on merging this PR? I am particularly interested in this feature (esp. the 3D filter part).

liruilong940607 commented:
I think this PR is just about cleaning up, add doc strings and update webpage. It would be nice to finish it up.

Not sure if @niujinshuchong still has time for it. But I might be able to find some cycles in the next two weeks to help with finishing up this PR.

niujinshuchong commented:
@liruilong940607 Yes, it just need to be clean up. I will find some time to continue. However, it would be great if you can help with it. Sorry for the late reply.
